extension/src/content.ts

```diff
@@ -1,40 +1,43 @@
-import { requestSail } from "./common";
+import { requestSail, SocketMessage } from "./common";
 
 const doConnection = (socketUrl: string, projectUrl: string, onMessage: (data: {
 	readonly type: "data" | "error";
 	readonly v: string;
-}) => void): Promise<WebSocket> => {
-	return new Promise<WebSocket>((resolve, reject) => {
-		const socket = new WebSocket(socketUrl);
-		socket.addEventListener("open", () => {
-			socket.send(JSON.stringify({
-				project: projectUrl,
-			}));
-
-			resolve(socket);
-		});
-		socket.addEventListener("close", (event) => {
-			reject(`socket closed: ${event.code}`);
-		});
-
-		socket.addEventListener("message", (event) => {
-			const data = JSON.parse(event.data);
-			if (!data) {
-				return;
-			}
-			const type = data.type;
-			const content = atob(data.v);
-
-			switch (type) {
-				case "data":
-				case "error":
-					onMessage({ type, v: content });
-					break;
-				default:
-					throw new Error("unknown message type: " + type);
-			}
-		});
-	});
+}) => void): Promise<chrome.runtime.Port> => {
+  return new Promise<chrome.runtime.Port>((resolve, reject) => {
+    const port = chrome.runtime.connect();
```

Comment:
Do `chrome.*` API calls still work in firefox?

Reply:
Yes. 

> Firefox also implements these APIs under the chrome namespace using callbacks. This allows code written for Chrome to run largely unchanged in Firefox for the APIs documented here. 

https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API#wikiArticle